File: arrays_and_strings/checkPermutation.cpp
// Given two strings, write a method to decide if one is a permutation of the other
#include <iostream>
#include <algorithm>
using namespace std;

bool checkPermuationSort(string s, string t) {
    // check if 2 strings have the same length
    if (s.length() != t.length()) {
        return false;
    }
    sort(s.begin(), s.end());
    sort(t.begin(), t.end());

    for (size_t i = 0; i < s.length(); i++) {
        if (s[i] != t[i]) return false;
    }
    return true;
};

int main() {
    string str1 = "hello";
    string str2 = "olhle";
    string str3 = "world";
    string str4 = "orlda";

    cout << "Using sort\n";
    if (checkPermuationSort(str1, str2)) {
        cout << str1 << " and " << str2 << " are permutation of each other\n";
    } else {
        cout << str1 << " and " << str2 << " are NOT permutation of each other\n";
    }
    if (checkPermuationSort(str3, str4)) {
        cout << str1 << " and " << str3 << " are permutation of each other\n";
    } else {
        cout << str1 << " and " << str4 << " are NOT permutation of each other\n";
    }

    cout << endl    ;
    return 0;
}

File: arrays_and_strings/python/check_permutation.py
"""given two string, write a method to decide if one is permutation of the other"""
import unittest
from collections import Counter

def check_permutation(str1, str2):
    """check permutation"""
    if len(str1) != len(str2):
        return False
    counter = Counter()
    for char in str1:
        counter[char] += 1
    for char in str2:
        if counter[char] == 0:
            return False
        counter[char] -= 1
    return True

class Test(unittest.TestCase):
    """docstring for Test."""
    dataT = (
        ('abcd', 'badc'),
        ('123456', '645321'),
        ('webc123', '213cbwe'),
    )
    dataF = (
        ('abcd', 'eadbl'),
        ('2315', '1234'),
        ('dbcw', 'bcwd5'),
    )

    def test_cp(self):
        """testing"""
        # true check
        for test_string in self.dataT:
            result = check_permutation(*test_string)
            self.assertTrue(result)
        # false check
        for test_string in self.dataF:
            result = check_permutation(*test_string)
            self.assertFalse(result)


if __name__ == '__main__':
    unittest.main()
